src/main/java/org/candlepin/audit/EventFactory.java

```diff
@@ -269,6 +269,60 @@ public Event bulkConsumerDeletion(String ownerKey, boolean anonymous, Collection
             .setEventData(Map.of("consumerUuids", consumerUuids));
     }
 
+    /**
+     * Generates a bulk consumer environment change event for the given owner and consumers.
+     *
+     * @param owner
+     *  The owner associated with the consumers that had their environments updated
+     *
+     * @param consumerUuids
+     *  The UUIDs of the consumers whose environment assignments changed
+     *
+     * @return an event representing the bulk environment change
+     */
+    public Event bulkConsumerEnvironmentChange(Owner owner, Collection<String> consumerUuids) {
+        if (owner == null) {
+            throw new IllegalArgumentException("owner is null");
+        }
+
+        return this.bulkConsumerEnvironmentChange(owner.getKey(), owner.getAnonymous(), consumerUuids);
+    }
+
+    /**
+     * Generates a bulk consumer environment change event for the given owner key and consumers.
+     *
+     * @param ownerKey
+     *  The key of the owner associated with the consumers
+     *
+     * @param anonymous
+     *  The anonymous state of the owner
+     *
+     * @param consumerUuids
+     *  The UUIDs of the consumers whose environment assignments changed
+     *
+     * @return an event representing the bulk environment change
+     */
+    public Event bulkConsumerEnvironmentChange(String ownerKey, boolean anonymous,
+        Collection<String> consumerUuids) {
+
+        if (ownerKey == null || ownerKey.isBlank()) {
+            throw new IllegalArgumentException("ownerKey is null or empty");
+        }
+
+        if (consumerUuids == null || consumerUuids.isEmpty()) {
+            throw new IllegalArgumentException("consumerUuids is null or empty");
+        }
+
+        Map<String, Object> eventData = new HashMap<>();
+        eventData.put("consumerUuids", consumerUuids);
+        eventData.put("operation", "setEnvironments");
```

Comment:
I'm not yet sure how I feel about this. By having an operation field in a more generic event, event processors have to always process this event and then inspect this field to determine whether or not it's an event of interest.

That's not necessarily the worst thing ever, but I don't believe we have much of that going on now, and in the cases we do, it's handled with the combination of TARGET+TYPE. This will introduce another level of event typing that isn't compatible with our existing methods (as right or wrong as they are).

Reply:
So I changed it to a level where we will be able to recognize what kind of operation it is at the target+type level.

Reply:
changed to BULK_MODIFICATION, becasue we do not need to know the operation